### PR TITLE
Add Azure network and VM clients with backoff retry

### PR DESCRIPTION
**Pull request by feiskyer**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

/kind feature
/sig cloud-provider
/area provider/azure
/priority important-soon

**What this PR does / why we need it**:

Continue of https://github.com/kubernetes-sigs/cloud-provider-azure/issues/247. This PR moves Azure network and VM clients to armclient-based, which supports backoff retry and honors Azure throttling headers.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Azure network and VM clients now suppress requests on throttling
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by andyzhangx** [comment]
would you squash your commits first?

**Comment by feiskyer** [comment]
> would you squash your commits first?

Let's keep the commit history, so it's easy to track the changes in the future.

File: staging/src/k8s.io/legacy-cloud-providers/azure/clients/subnetclient/doc.go
Comment (feiskyer) [reply staging/src/k8s.io/legacy-cloud-providers/azure/clients/subnetclient/doc.go]: this allows kubernetes components to be built without any cloud providers

**Comment by liggitt** [comment]
modules change is fine, tag me for approval when the rest of the PR has lgtm

**Comment by feiskyer** [comment]
pull-kubernetes-e2e-azure-disk is failing with error: `"message": "The property 'dataDisk.createOption' cannot have the value 'Empty' when property 'dataDisk.managedDisk.id' is set."`

**Comment by andyzhangx** [comment]
> pull-kubernetes-e2e-azure-disk is failing with error: `"message": "The property 'dataDisk.createOption' cannot have the value 'Empty' when property 'dataDisk.managedDisk.id' is set."`

the original 'dataDisk.createOption' logic is here:
https://github.com/kubernetes/kubernetes/blob/61d36e4a43b831f960b190b81d371cb33b5f20d1/staging/src/k8s.io/legacy-cloud-providers/azure/azure_controller_common.go#L336-L370

I think there is sth wrong with the vmas client code, could you remove that commit and merge other code first? @feiskyer

**Comment by feiskyer** [comment]
> I think there is sth wrong with the vmas client code, could you remove that commit and merge other code first?

Still not sure why the data internal details are touched here. Let's troubleshoot what's wrong here.

/test pull-kubernetes-e2e-azure-disk

**Comment by andyzhangx** [comment]
I guess there is sth wrong with disk driver on vmas on 1.18, will check

**Comment by feiskyer** [comment]
#87210 would fix the disk e2e failures, would rebase after that PR gets merged.

**Comment by feiskyer** [comment]
PR #87210 gets merged and rebased the PR. Rerun the tests now.

/test pull-kubernetes-e2e-aks-engine-azure
/test pull-kubernetes-e2e-azure-disk
/test pull-kubernetes-e2e-azure-disk-vmss

**Comment by feiskyer** [comment]
test passed. ping @andyzhangx @liggitt for approval

**Review by andyzhangx — APPROVED** [review]
/lgtm

**Pull request merged**
